action/protocol/staking/protocol.go

```diff
@@ -178,6 +179,27 @@ func (p *Protocol) Start(ctx context.Context) error {
 			return err
 		}
 	}
+
+	// load the candidate center to StateReader
+	srCtx := protocol.MustGetStateReaderCtx(ctx)
+	return srCtx.SR.Load(protocolID, p.inMemCandidates)
+}
+
+// CreatePreStates creates states before Handle
+func (p *Protocol) CreatePreStates(ctx context.Context, sm protocol.StateManager) error {
+	// unload candidate center
+	v, err := sm.Unload(protocolID)
+	if err != nil {
+		return err
+	}
+
+	center, ok := v.(*CandidateCenter)
+	if !ok {
+		return ErrUnload
+	}
+
+	p.inMemCandidates = nil
+	p.inMemCandidates = center.Clone()
```

Comment:
Moreover, inMemCandidates is not a variable of protocol, but something stored in sr.

Reply:
yes, it is stored in SR, we call Unload() above to read it from SR

Reply:
i think SR should only provide a means for protocol to store their data in general, that is why the data is defined as interface{}

it is up to each protocol to interpret the actual format of that data, and here it is *CandidateCenter

Reply:
My previous explanation may not be clear. Let me rephrase it:
0. CreatePreStates is not used for this purpose, it just happens to be called before validating or committing a block
1. CandidateCenter is cloned when workingset is created, it is a variable stored in a working set (I can understand that we may only have one workingset at a time, that in current implementation, it behaves the same, but the way of cloning it is not appropriate)
2. Whenever we need to use CandidateCenter, it should be derived from sm. It is the same way as what you do here, but it is per sm. e.g., in function handleUnstake:
`func (p *Protocol) Hanle(ctx context.Context, act *action.Unstake, sm protocol.StateManager) (*action.Receipt, error) {`
we can convert the "sm" into a CandidateCenter, and then pass it to the sub handlers, that the sub handlers do not need to know the existence of "sm".
3. With the above two points, we don't really need a struct level variable inMemCandidates

Reply:
well, isn't CreatePreStates() called before ws.runActions() to prepare the protocol for upcoming Validate() and Handle() calls? so i think it is the place for this purpose.

i know what you mean, and i think we have an issue/bug here. The block may have more than 1 staking actions, say voting for 3 diff candidates, we cannot have each handler read the (same) CandidateCenter, and push the dirty CandidateCenter (with 1 change) back to sm.

we still need a protocol-level inMemCandidates to collect all 3 changes, and push the dirty CandidateCenter (with 3 changes), after ws.runActions().

as we discussed earlier, CreatePostSystemActions() is only called in MintNewBlock(). So for this Dock idea to work, we'll need to add a similar interface like PostHandle(), and call that after ws.runActions(). Or we fallback to read all candidates from DB and construct the CandidateCenter on-the-fly in CreatePreStates()

Reply:
thx, your comments opened more thoughts and insights

i structured it a bit such that this loading/syncing of CandidateCenter would only be needed when the protocol is actually called, that is, if the block contains staking actions. This avoids the unnecessary preparation of the CandidateCenter for irrelevant blocks.

pls check the latest commit in this PR